### PR TITLE
Feat: Organisation allfeat

### PR DESCRIPTION
**Pull request by IgweWisdomJaminel**
# API Endpoints for Organization Roles and Permissions

### Create a new organization
- **Endpoint:** `POST /api/v1/organizations`
- **Description:** Creates a new organization.

### Update an organization
- **Endpoint:** `PATCH /api/v1/organizations/:org_id`
- **Description:** Updates details of an existing organization.

### Get organization details
- **Endpoint:** `GET /api/v1/organizations/:org_id`
- **Description:** Fetches details of a specific organization.

### Delete an organization
- **Endpoint:** `DELETE /api/v1/organizations/:org_id`
- **Description:** Deletes a specific organization.

### Get users in an organization
- **Endpoint:** `GET /api/v1/organizations/{org_id}/users`
- **Description:** Retrieves all users within a specific organization.

### Invite a user to an organization
- **Endpoint:** `POST /api/v1/organizations/:org_id/send-invite`
- **Description:** Sends an invitation to a user to join the organization via email.

### Remove a user from an organization
- **Endpoint:** `DELETE /api/v1/organizations/{org_id}/users/{user_id}`
- **Description:** Removes a specific user from the organization.

### Accept an invitation
- **Endpoint:** `POST /api/v1/organizations/accept-invite?token=`
- **Description:** Allows a user to accept an invitation to join an organization.

### Create a role in an organization
- **Endpoint:** `POST /api/v1/organizations/{org_id}/roles`
- **Description:** Creates a new role within a specific organization.

### Get all roles in an organization
- **Endpoint:** `GET /api/v1/organizations/{org_id}/roles`
- **Description:** Retrieves all roles within a specific organization.

### Get a single role in an organization
- **Endpoint:** `GET /api/v1/organizations/{org_id}/roles/{role_id}`
- **Description:** Retrieves details of a specific role within an organization.

### Update a role in an organization
- **Endpoint:** `PATCH /api/v1/organizations/{org_id}/roles/{role_id}`
- **Description:** Updates details of a specific role within an organization.

### Soft delete a role in an organization
- **Endpoint:** `DELETE /api/v1/organizations/{org_id}/roles/{role_id}`
- **Description:** Soft deletes a specific role within an organization.

### Update role permissions in an organization
- **Endpoint:** `PATCH /api/v1/organizations/{org_id}/{role_id}/permissions`
- **Description:** Updates permissions of a specific role within an organization.

### Export members of an organization
- **Endpoint:** `GET /api/v1/members/export`
- **Description:** Exports all members of an organization.

## How should this be manually tested

- Ensure you have the latest code from the `main` branch.
- Ensure your environment variables for database and authentication are set correctly.

### Steps to Test
1. **Clone the repository**: `git clone <repository-url>`
2. **Checkout to the new branch**: `git checkout <your-branch-name>`
3. **Compile the application**: `javac application.java`
4. **Run the application**: `java application.java start`
5. **Authenticate and obtain a valid token**: Use the login endpoint to get a token.
6. **Issue a DELETE request** to the endpoint:
   - Use a tool like Postman or curl:
     ```
     DELETE /api/v1/organizations/<org_id>
     Content-Type: application/json
     Authorization: Bearer <token>
     ```
     - Replace `<org_id>` with a valid organization ID and `<token>` with your authentication token.
7. **Verify the responses**:
   - For a valid request, expect a **204 No Content** status.
   - For an invalid organization ID format, expect a **400 Bad Request** status.
   - For a non-existent organization ID, expect a **404 Not Found** status.
   - For unauthorized deletion attempt, expect a **401 Unauthorized** status.

### Expected Outcomes
- **204 No Content**: The organization is marked as deleted.
- **400 Bad Request**: Invalid organization ID format.
- **404 Not Found**: Invalid organization ID.
- **401 Unauthorized**: User not authorized to delete this organization.

## Checklist

- [x] Developed server-side logic to handle organization deletion requests.
- [x] Validated and sanitized incoming organization ID data.
- [x] Validated that the user deleting the organization is the owner.
- [x] Changed the specified organization's `deleted` attribute to `true` in the database upon successful validation.
- [x] Maintained data integrity by keeping all related data of the specified organization intact.
- [x] Wrote unit tests to ensure the organization deletion endpoint validates input correctly and removes the organization from the system.
- [x] Tested various scenarios for submitting the organization ID (e.g., valid ID, non-existent ID, malformed ID, etc.).
- [x] Verified that all unit tests pass.
- [x] Ensured the code follows the project's style guidelines.
- [x] Removed all console logs and comments.
- [x] Updated the documentation as necessary.
- [x] Informed relevant stakeholders of the changes.

## Reference issue
[https://github.com/hngprojects/hng_boilerplate_java_web/issues/18](https://github.com/hngprojects/hng_boilerplate_java_web/issues/18)

[https://github.com/hngprojects/hng_boilerplate_java_web/issues/21]
(https://github.com/hngprojects/hng_boilerplate_java_web/issues/21)

[https://github.com/hngprojects/hng_boilerplate_java_web/issues/115]
(https://github.com/hngprojects/hng_boilerplate_java_web/issues/115)

### CODE REVIEW

**Review by lurldgbodex — CHANGES_REQUESTED** [review]
remove the application.properties file

**Review by lurldgbodex — CHANGES_REQUESTED** [review]
I can't seem to find the migration files for the entities your created and altered

**Comment by IgweWisdomJaminel** [comment]
Sorry made many modifications and tested on my local machine because some merged feature were not working properly I have corrected it now

File: src/main/resources/application.properties
Comment (lurldgbodex) [inline src/main/resources/application.properties:1]: you need to delete the application.prorperties file

File: src/main/resources/db/migration/V22__create_Invitation_table.sql
Comment (lurldgbodex) [inline src/main/resources/db/migration/V22__create_Invitation_table.sql:1]: update the migration version. check the repo to see the latest version

File: src/main/resources/db/migration/V9__update_organisations.sql
Comment (lurldgbodex) [inline src/main/resources/db/migration/V9__update_organisations.sql:1]: do not edit existigng migration file. create a new migration file and update the table

Comment (IgweWisdomJaminel) [reply src/main/resources/db/migration/V9__update_organisations.sql:1]: made all changes already boss

**Comment by lurldgbodex** [comment]
you did not change the migration version. I'm still seeing V22__, V23___ etc.Also use 1 file for all your new migrations

**Review by lurldgbodex — APPROVED** [review]

**Pull request closed without merging**